turbopack/crates/turbopack-ecmascript/src/lib.rs

```diff
@@ -1304,92 +2005,544 @@ fn process_content_with_code_gens(
+        match self {
+            CodeGenResultSourceMap::Single { .. } => f.start_pos <= raw_pos && raw_pos < f.end_pos,
+            CodeGenResultSourceMap::ScopeHoisting { .. } => {
+                // let (module, pos) = CodeGenResultComments::decode_bytepos(*modules_header_width,
+                // pos);
+
+                // TODO optimize this, unfortunately, `SourceFile` doesn't know which `module` it
+                // belongs from.
+                false
+            }
+        }
     }
 
-    fn add_leading_comments(&self, pos: BytePos, comments: Vec<Comment>) {
-        self.mutable.add_leading_comments(pos, comments);
+    fn map_raw_pos(&self, pos: BytePos) -> BytePos {
+        match self {
+            CodeGenResultSourceMap::Single { .. } => pos,
+            CodeGenResultSourceMap::ScopeHoisting {
+                modules_header_width,
+                ..
+            } => CodeGenResultComments::decode_bytepos(*modules_header_width, pos).1,
+        }
+    }
+}
+
+impl SourceMapper for CodeGenResultSourceMap {
+    fn lookup_char_pos(&self, pos: BytePos) -> Loc {
+        match self {
+            CodeGenResultSourceMap::Single { source_map } => source_map.lookup_char_pos(pos),
+            CodeGenResultSourceMap::ScopeHoisting {
+                modules_header_width,
+                source_maps,
+            } => {
+                let (module, pos) =
+                    CodeGenResultComments::decode_bytepos(*modules_header_width, pos);
+                source_maps[module].lookup_char_pos(pos)
+            }
+        }
+    }
+    fn span_to_lines(&self, sp: Span) -> FileLinesResult {
+        match self {
+            CodeGenResultSourceMap::Single { source_map } => source_map.span_to_lines(sp),
+            CodeGenResultSourceMap::ScopeHoisting {
+                modules_header_width,
+                source_maps,
+            } => {
+                let (module, lo) =
+                    CodeGenResultComments::decode_bytepos(*modules_header_width, sp.lo);
+                source_maps[module].span_to_lines(Span {
+                    lo,
+                    hi: CodeGenResultComments::decode_bytepos(*modules_header_width, sp.hi).1,
+                })
+            }
+        }
+    }
+    fn span_to_string(&self, sp: Span) -> String {
+        match self {
+            CodeGenResultSourceMap::Single { source_map } => source_map.span_to_string(sp),
+            CodeGenResultSourceMap::ScopeHoisting {
+                modules_header_width,
+                source_maps,
+            } => {
+                let (module, lo) =
+                    CodeGenResultComments::decode_bytepos(*modules_header_width, sp.lo);
+                source_maps[module].span_to_string(Span {
+                    lo,
+                    hi: CodeGenResultComments::decode_bytepos(*modules_header_width, sp.hi).1,
+                })
+            }
+        }
+    }
+    fn span_to_filename(&self, sp: Span) -> Arc<FileName> {
+        match self {
```

Comment:
Can we have some method to get an instance of `&swc_common::SourceMap` from `CodeGenResultSourceMap` and use it instead?

Reply:
If swc had such an API, yes that would be possible.
Would that have concrete benefits or just mean having to write less code?